test/unit/app/utils/test_kenya_state_machine.py

```diff
@@ -154,15 +163,72 @@ def test_change_language(mocker, test_client, init_database, session_factory, us
     state_machine.send_sms.assert_called_with("language_change_sms")
 
 
-# TODO: need pin authorisation implemented to do this
 def test_opt_out_of_marketplace(mocker, test_client, init_database):
     session = opt_out_of_market_place_pin_authorization_state()
     user = standard_user()
     user.phone = phone()
+    assert len(user.custom_attributes.all()) == 0
     state_machine = KenyaUssdStateMachine(session, user)
     state_machine.send_sms = mocker.MagicMock()
 
-    #state_machine.feed_char("0000")
-    #assert state_machine.state == "complete"
-    #assert user.custom_attributes.market_enabled == False
-    #state_machine.send_sms.assert_called_with("opt_out_of_market_place_sms")
+    state_machine.feed_char("0000")
+    assert state_machine.state == "complete"
+    market_enabled = list(filter(lambda x: x.name == 'market_enabled', user.custom_attributes.all()))
+    assert market_enabled[0].value == 'false'
+    state_machine.send_sms.assert_called_with("opt_out_of_market_place_sms")
+
+
+# TODO: when add the other things requiring pin auth, also test the failure states here
+@pytest.mark.parametrize("session_factory, user_factory, user_input, expected, before_failed_pin_attempts, after_failed_pin_attempts",
```

Comment:
finished opt_out_of_marketplace test, removed redundant authorize_pin test, removed separate helpers for various numbers of failed_pin_attempts, instead just instantiated them in the tests here